src/Language/PureScript/Kinds.hs

```diff
@@ -1,109 +1,164 @@
       throwCustomError (T.append "Unrecognised tag: " other)
 
   where
   -- The following are copied from Environment and reimplemented to avoid
   -- circular dependencies.
-  primName :: Text -> Qualified (ProperName a)
+  primName :: Text -> Qualified (ProperName b)
   primName = Qualified (Just $ ModuleName [ProperName C.prim]) . ProperName
 
-  primKind :: Text -> Kind
-  primKind = NamedKind . primName
+  primKind = flip NamedKind . primName
 
   kindType = primKind C.typ
   kindSymbol = primKind C.symbol
 
-instance A.FromJSON Kind where
-  parseJSON = toAesonParser id kindFromJSON
+instance {-# OVERLAPPING #-} A.FromJSON (Kind SourceAnn) where
+  parseJSON = toAesonParser id (kindFromJSON (pure NullSourceAnn) fromAesonParser)
+
+instance {-# OVERLAPPING #-} A.FromJSON a => A.FromJSON (Kind a) where
+  parseJSON = toAesonParser id (kindFromJSON (fail "Invalid annotation") fromAesonParser)
 
-everywhereOnKinds :: (Kind -> Kind) -> Kind -> Kind
+everywhereOnKinds :: (Kind a -> Kind a) -> Kind a -> Kind a
 everywhereOnKinds f = go
   where
-  go (Row k1) = f (Row (go k1))
-  go (FunKind k1 k2) = f (FunKind (go k1) (go k2))
+  go (Row ann k1) = f (Row ann (go k1))
+  go (FunKind ann k1 k2) = f (FunKind ann (go k1) (go k2))
   go other = f other
 
-everywhereOnKindsM :: Monad m => (Kind -> m Kind) -> Kind -> m Kind
+everywhereOnKindsM :: Monad m => (Kind a -> m (Kind a)) -> Kind a -> m (Kind a)
 everywhereOnKindsM f = go
   where
-  go (Row k1) = (Row <$> go k1) >>= f
-  go (FunKind k1 k2) = (FunKind <$> go k1 <*> go k2) >>= f
+  go (Row ann k1) = (Row ann <$> go k1) >>= f
+  go (FunKind ann k1 k2) = (FunKind ann <$> go k1 <*> go k2) >>= f
   go other = f other
 
-everythingOnKinds :: (r -> r -> r) -> (Kind -> r) -> Kind -> r
+everythingOnKinds :: (r -> r -> r) -> (Kind a -> r) -> Kind a -> r
 everythingOnKinds (<>.) f = go
   where
-  go k@(Row k1) = f k <>. go k1
-  go k@(FunKind k1 k2) = f k <>. go k1 <>. go k2
+  go k@(Row _ k1) = f k <>. go k1
+  go k@(FunKind _ k1 k2) = f k <>. go k1 <>. go k2
   go other = f other
+
+annotationForKind :: Kind a -> a
+annotationForKind (KUnknown a _) = a
+annotationForKind (Row a _) = a
+annotationForKind (FunKind a _ _) = a
+annotationForKind (NamedKind a _) = a
+
+instance Eq (Kind a) where
+  (==) = eqKind
+
+instance Ord (Kind a) where
+  compare = compareKind
+
+eqKind :: Kind a -> Kind b -> Bool
+eqKind (KUnknown _ a) (KUnknown _ a') = a == a'
+eqKind (Row _ a) (Row _ a') = eqKind a a'
+eqKind (FunKind _ a b) (FunKind _ a' b') = eqKind a a' && eqKind b b'
+eqKind (NamedKind _ a) (NamedKind _ a') = a == a'
+eqKind _ _ = False
+
+compareKind :: Kind a -> Kind b -> Ordering
+compareKind (KUnknown _ a) (KUnknown _ a') = compare a a'
+compareKind (KUnknown {}) _ = LT
+
+compareKind (Row _ a) (Row _ a') = compareKind a a'
+compareKind (Row {}) _ = LT
+
+compareKind (FunKind _ a b) (FunKind _ a' b') = compareKind a b <> compareKind a' b'
+compareKind (FunKind {}) _ = LT
+compareKind _ (FunKind {}) = GT
```

Comment:
Is this necessary? Presumably if we reach this case, the first argument must be a `NamedKind`, and therefore we can let it fall through and hit the very last case? Or, alternatively, if we're following the style of `compareType`, should we have a case like this for the other constructors too?

Reply:
You are right, this looks fishy.